Deploy MetalLB and nginx-ingress

NicolasT opened the pull request:
**Component**: salt, kubernetes, metallb, ingress, nginx-ingress

**Summary**: deploy MetalLB and `nginx-ingress` as part of the cluster.

**Acceptance criteria**: 
- Deploy the cluster
- Find the `nginx-ingress-controller` `LoadBalancer` service (now managed by MetalLB):

```
$ kubectl -n nginx-ingress-system get svc
...
nginx-ingress-system   nginx-ingress-controller        LoadBalancer   10.101.134.174   172.21.254.47   80:30107/TCP,443:31177/TCP   6m
...
```

- Ensure the `metalk8s-ui` `Ingress` exists:

```
$  kubectl -n kube-system get ingress metalk8s-ui
NAME          HOSTS      ADDRESS   PORTS     AGE
metalk8s-ui   metalk8s             80        3h
```

- Either set up `/etc/hosts` for `metalk8s` to point to the LB address, or use `curl` and name resolution override to validate access to the admin UI service through the LB address using

```
$  curl http://metalk8s --resolve 'metalk8s:80:172.21.254.47'
...
```

Closes: #779
Closes: #791

bert-e commented:
# Hello nicolast,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/779-metallb` into
`development/2.0`
* [w/2.1/feature/779-metallb](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/779-metallb) will be merged into `development/2.1` 
* [w/2.2/feature/779-metallb](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/779-metallb) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @jbertran

* @monpote

* @alxf

* @gdemonet

* @lucieleonard

* @zempashi

* @carlito-scality

* @alexandre-allard-scality

* @nootal

* @nicolast

* @thomasdanan

* @tcarmet

* @slaperche-scality

* @teddyandrieux

* @sayf-eddine-scality

* @ramanduh

NicolasT committed:
charts: import metallb-0.9.5

```
$ helm fetch -d charts --untar stable/metallb
```

slaperche-scality commented on buildchain/buildchain/image.py:
```diff
@@ -249,6 +249,24 @@ def show() -> str:
+        registry=constants.METALLB_REGISTRY,
+        remote_name='controller',
+        name='metallb-controller',
+        version='v0.7.3',
```
Minor: we could have a constant `LLB_VERSION : str = 'v0.7.3'` so that when we bump the version we have a single replace instead of several.

NicolasT committed:
charts: quote image name in MetalLB templates

During `helm template`, we splice a Jinja macro into the container image
repository value. When not quoted properly, later parsing of the
resulting YAML to fixup namespace information and such fails, because
the result is (prior to Jinja rendering) not valid YAML.

Simply quoting the YAML string value properly solves this.

NicolasT committed:
salt: add `build_image_repo` macro

NicolasT committed:
salt: deploy MetalLB from chart

This is currently a hack, where the `build.sh` script is used to
generate `salt/metalk8s/addons/metallb/deployed/chart.sls`, which should
be part of the build-system instead of being checked into VCS.

bert-e commented:
# Conflict




There is a conflict between your branch `feature/779-metallb` and the
destination branch `development/2.0`.

Please resolve the conflict on **the feature branch** (`feature/779-metallb`).

```bash
 $ git fetch
 $ git checkout origin/feature/779-metallb
 $ git merge origin/development/2.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:feature/779-metallb
```

NicolasT committed:
charts: import `nginx-ingress` 1.6.13

```
$ helm fetch -d charts --untar stable/nginx-ingress
```

NicolasT committed:
charts: fix image name quotes in `nginx-ingress` templates

We use the Helm chart as a template rendering into Jinja templates. To
splice in variables that contain quotes, we need to inverse the
quotations.

See: 79e45b695627bf0c4f9ba06eaba9d1924cc6dd22

NicolasT committed:
salt: deploy `nginx-ingress`

Fixes: #791
See: https://github.com/scality/metalk8s/issues/791

NicolasT committed:
addons: create an `Ingress` for `metalk8s-ui`

NicolasT committed:
salt: fix `Service` port handling in `metalk8s_kubernetes`

Given our use of YAML manifests, we should support 'original' Kubernetes
objects and their attributes, not (only) the mangled Python'esque
versions.

See: fe94d24bbc5e633c982b3158419865fbd80dfbca

bert-e commented:
# History mismatch



Merge commit #290d349f07125c66f83e42a3f876002cd10399e5 on the integration branch
`w/2.1/feature/779-metallb` is merging a branch which is neither the current
branch `feature/779-metallb` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/779-metallb` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT commented:
At least two outstanding TODOs:

- [ ] Currently, the Vagrant VIP range is hard-coded in the MetalLB configuration, which should of course be configurable
- [ ] The Helm charts are rendered through a script and imported into VCS as-is. This should instead be done by the buildchain as a new kind of target, taking the chart tree and `values.yaml` as inputs.

NicolasT committed:
packages: use `containerd` from Fedora 30

The `nginx-ingress-controller` container relies on filesystem extended
attributes to allow permissions to the `nginx` binary and temporary
storage ACLs. There's a bug in `containerd` 1.2.1 which causes such
xattrs not to be properly unpacked from container image layers. To
work-around this, we need a newer version, which is not packaged for
CentOS/EPEL. However, since `containerd` is a pure Golang static binary,
we can use a package from Fedora.

See: https://github.com/kubernetes/ingress-nginx/issues/4061#issuecomment-496294198
See: https://github.com/containerd/containerd/issues/2942

NicolasT commented:
@sayf-eddine-scality is up-to-date about this one.

NicolasT commented:
`nginx-ingress` is handled in #1445, and the use of MetalLB is unclear (cfr. #1788), and integration of the chart would likely need to be redone anyway.

Hence, closing this one.
